### ISSUE TITLE
Does punq support scopes / instantiation methods?

### ISSUE THREAD

**Issue by jbcpollak** [bobthemighty/punq#26]
A lot of DI frameworks support specifying scope of registered objects. For example, if a new instance should be created each time a registered object is resolved, or if only one should be instantiated. 

For example, I'd like to be able to do:

`container.register(Type, MyImplementation, mode=singleton)`

then when I do this:

```
t1 = container.resolve(Type)
t2 = container.resolve(Type)
```

t1 and t2 would be the same instance.

This is slightly different than using the `.register(Type, instance=my_instance)` because in the first example I don't have to create the instance manually.

**Comment by bobthemighty** [comment]
At the moment, not explicitly.

My intention with punq was to build a simple DI container that covered the 95 percentile use case. It's possible to cover singletone lifecycles with a factory function, though, that returns the constructed instance if it already exists, otherwise creates a new one.

**Comment by jbcpollak** [comment]
Ah, just to be clear does that mean if we resolve the same type twice, we'll get two separate instances of that type?

I typically find I want to do the opposite, so I'll need to keep that in mind. Would you accept a PR to control that?

**Comment by jbcpollak** [comment]
It looks to me like Punq's default seems to be "singleton scope" because I see in `_build_impl()` it cache's the implementation:

```
context[registration.service] = result
```

and in `_resolve_impl()` its returned:

```
        if context.has_cached(service_key):
            return context[service_key]
```

Spring calls the alternative "prototype" scope, which would return a new instance for each call to `.resolve()`. Personally I don't like that mode, but some DI frameworks default to that.

I was thinking of adding a Scope enumeration and a `register(scope=Scope.singleton)` function parameter, but since it already defaults to that mode I personally don't have much need for prototype mode, I might delay on doing it.

My plan was to persist the scope in the Registration object, then use that to control if caching was done in `_build_impl()` or not.

In the meantime I'd like to add a unit test to ensure the singleton scope is ensured by default. In the future that could be extended to support prototype mode as well.

**Comment by jbcpollak** [comment]
It looks like I misunderstood in my message above. I tried to implement as I described and it didn't work correctly. Please see the PR here (which is _not_ working):

https://github.com/bobthemighty/punq/pull/27

I'd appreciate feedback on two points:
* does this API make sense? Would you accept a PR for this feature?
* how do I make it work? I think I need to cache the context object on this line:
`        context = self.registrations.build_context(service_key)`
instead of playing with the context's cache itself. If that's not right, please let me know.

TY.

**Comment by bobthemighty** [comment]
The context object's responsibility is to use the same instance of an object within a _single_ resolution. If, for example, you wanted to resolve a set of objects that all required a repository, or configuration or something, we'd use a single instance to satisfy each constructor.

I'm definitely opposed to having a singleton default, because I think it's often a poor design choice. I had a lot of fun and games with singleton registrations back in my .NET days, and the community were generally in favour of transient resolutions where possible. Services, which are the usual target for DI, ought to be stateless, so for the 99th percentile case, it oughtn't to matter whether the resolved item is a singleton or transient instance. Is there a particular reason you want singletons?

Given that a singleton is achievable with a factory function, or by passing the instance as a kwarg, I'd be reluctant to make wide changes to the API to support them. That said, I'm not opposed to a PR for the functionality so long as transient instances are the default, and I don't have to care about scopes unless I want something else.

**Comment by jbcpollak** [comment]
> The context object's responsibility is to use the same instance of an object within a single resolution. If, for example, you wanted to resolve a set of objects that all required a repository, or configuration or something, we'd use a single instance to satisfy each constructor.

That makes sense. Does it make sense to cache the context in `resolve()` to support singletons, or do you think the integration needs to go deeper? I think if the repository (in your example above) is flagged as a singleton, it needs to be in its own context, so that two transient services that depend on the repository would get the same instance. That makes me think something like a global store of scope-aware contexts is needed, at least for singletons.

> I'm definitely opposed to having a singleton default, because I think it's often a poor design choice. I had a lot of fun and games with singleton registrations back in my .NET days, and the community were generally in favour of transient resolutions where possible. Services, which are the usual target for DI, ought to be stateless, so for the 99th percentile case, it oughtn't to matter whether the resolved item is a singleton or transient instance. Is there a particular reason you want singletons?

All good points, I agree with keeping transient as the default, I only made singleton the default because I mis-read the code and thought it already was. I wouldn't want to change the default behavior. I've switched it back locally and what will be part of the PR if I get it working.

I agree Services should be stateless, although for me that argues since they are stateless there is no harm reusing them as needed (when they are properly written, which is admittedly hard to ensure). 

There are a few reasons why I like/need singleton services: One, since they are stateless, as a memory optimization technique and to reduce the time it takes to resolve/instantiate services, being able to reuse them is helpful. Two, I'm often integrating with hardware or other systems that don't tolerate multiple connections. So while the service itself may be stateless, it is representing state or has threads running that communicate with other devices and we don't want multiple instances of the service running.

Side note: do you prefer the term transient to prototype? I don't feel strongly one way or the other.

> Given that a singleton is achievable with a factory function, or by passing the instance as a kwarg, I'd be reluctant to make wide changes to the API to support them. That said, I'm not opposed to a PR for the functionality so long as transient instances are the default, and I don't have to care about scopes unless I want something else.

The problem with an instance is its clunky to do the instantiation outside of punq, especially if you don't want to worry about the type you are creating for the interface. I still want to be able to leverage punq's instantiation abilities.

I can't think of a way to leverage factories to make this cleaner, but if you have a suggestion I'm all ears.

**Comment by bobthemighty** [comment]
Closed by #27
